gbn/gbn_server.go

```diff
@@ -111,6 +112,24 @@ func (g *GoBackNConn) serverHandshake() error { // nolint:gocyclo
 
 		switch msg.(type) {
 		case *PacketSYN:
+
+		case *PacketSYNACK, *PacketData:
+			// If we receive a SYNACK or DATA packet after we have
+			// restarted the handshake, we can be sure that the
+			// client has received our SYN and has completed the
+			// handshake. We can therefore complete the handshake
+			// ourselves.
```

Comment:
> we can be sure that the client has received our SYN

Hmmm no I dont think that is true.

What if the client is still on a previous connection resending Data with an Old `N` value? ie, it is not on a handshake at all? 

With the handshake, we need to not make any assumptions - we should be very sure that they are on the same page with the same N value. 

A couple of handshake retries is not a bad thing I think. 

let me know if I am missing something 🙏

Reply:
Hmm I don't think that could become an issue with these changes?
Since the handshake process is:
1. Client: SYN
2. Server: SYN
3. Client: SYNACK

Since this newly added `if` case that allows the server to consider the client to have completed the handshake, only happens when the server has already received the client's SYN, we can be sure that the client isn't resending Data with an old `N`, as that can't happen after the client has sent the SYN. Or am I missing something?

What happened before is that the client could move on from the handshake, while the server would be stuck in it.
Just for context, I'm attaching the `itest` CI workflow that failed due to that:
https://github.com/lightninglabs/lightning-node-connect/actions/runs/7446471814/job/20256804763

Let me know though my understanding is incorrect, and this indeed could be an issue!

Reply:
ok cool 👍